analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.NullableContext.cs

```diff
@@ -0,0 +1,48 @@
+﻿#nullable enable
+
+using System;
+
+class NullForgivingOperator
+{
+    void Basics(int? i)
+    {
+        _ = i!.Value;               // Compliant, unknown
+        i = SomeMethod();
+        _ = i!.Value;               // Compliant, unknown
+
+        i = null;
+        _ = i!.Value;               // Noncompliant, empty
```

Comment:
What is our strategy here? With the "possible null reference" rule we honored the user's wish to overrule us. If this should be the case here too, then this needs to be marked as FP.
What is the "empty" comment referring to? Please use "null" here and below instead.

Reply:
My understanding, deriving from a discussion we had during one of our sync meeting, is the following.

In the rule checking about possible null references, we trigger an issue every time we find a dereference of something which can be `null`. If we don't know for certain (i.e. unknown value), we trigger, because the possibility of dereference is there.

In this rule, however, we only trigger when we know for certain that there is a path where the variable is `null`. In the case above, for example, `i` is assigned to be `null` right before value accessing, so we know for certain that an `InvalidOperationException` will be produced as a result of calling `Value` on `i`, no matter if the null forgiving operator is used or not.

So we basically ignore and ask the user to protect against the scenario where the variable surely get a `null`.

An objection to this line of reasoning may be that a simple check of `HasValue` would trigger the rule:
```cs
void Method(int? i)
{
    if (i.HasValue) { }
    _ = i.Value; // Noncompliant, there's a path where i is empty
}
```
whereas
```
void Method(int? i)
{
    _ = i.Value; // Compliant, unknown
}
```

My understanding (and I would ask @pavel-mikula-sonarsource to back up or dismiss the reasoning) is that if the user has made a previous call to `i.HasValue`, he has somewhat acknowledged the fact that `i` can be empty. So we then ask him to be consistent with that observation, and protect his access to `i.Value` with a check `i.HasValue` or `i != null`.

Concerning the use of `, empty` after `Noncompliant`: it refers to the nullable, which is empty. I have used the word empty, rather than `null`, because the rule is called `EmptyNullableValueAccess` and also because I find misleading saying that a `struct` can be `null`. It's rather the opposite: `null` is converted by the compiler into a `Nullable<T>` struct with `hasValue == false`.

Reply:
We should honor the null forgiving operator. In some cases, our engine is wrong, and the user should be able to make the rule shut up.

Reply:
To correct @antonioaversa understanding: Both S3655 and S2259 behave the same in principle. If user indicated possible null (for example by asking `nullable.HasValue` or `arg is null` on the other rule), we assume possible null and will raise. S2259 does not raise if we don't know to reduce noise.

As `!` means "trust me, it works (and I won't blame you)", I'm fine with keeping the behavior symmetrical and support bang here too.

Reply:
Great, thanks for the explanation. Support of this will come in a dedicated PR: https://github.com/SonarSource/sonar-dotnet/pull/7015